c10/macros/Macros.h

```diff
@@ -362,7 +363,13 @@ extern SYCL_EXTERNAL void __assert_fail(
 #else // __SYCL_DEVICE_ONLY__
 #if (defined(__CUDA_ARCH__) && !(defined(__clang__) && defined(__CUDA__)))
 __host__ __device__
-#endif // __CUDA_ARCH__
+#elif defined(__HIP_ARCH__) || defined(__HIP__)
+__device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(
+    const char* assertion,
+    const char* file,
+    unsigned int line,
+    const char* function);
+#endif
```

Comment:
Thanks for your PR. It's really hard to understand how this will be expanded. It seems that
```cxx
# CUDA: one declaration
__host__ __device__ void __assert_fail(...) throw() __attribute__((__noreturn__));

# HIP/ROCm: two declarations
__device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(...);
void __assert_fail(...) throw() __attribute__((__noreturn__));
```

Suggestions:
1. Would you mind adding a comment to the code?
2. Can you clean up this `#ifdef` logic? Perhaps:
```cxx
#if CUDA
__host__ __device__
#endif
void __assert_fail(...) throw() __attribute__((__noreturn__));

#if HIP
// A comment on a reason why it is needed
__device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(...);
#endif
```

Reply:
Updated the code.